patina_dxe_core/src/image.rs

```diff
@@ -375,43 +382,38 @@ impl PrivateImageData {
         result
     }
 
-    /// Calculates and sets the file name for this image, if appropriate
+    /// Sets both the file path and file name for this image.
+    ///
+    /// The file name is a part of the loaded_image protocol, and is the remaining portion of the device path after
+    /// the parent handle's device path (if there is a valid parent handle).
     ///
-    /// If the device_handle is valid, determines the file name from the device path instance installed on the device_handle.
-    /// If the device_handle is invalid, uses the provided file name directly.
-    fn set_file_name(&mut self, file_path: *mut efi::protocols::device_path::Protocol) -> Result<(), EfiError> {
-        let fixed_file_path = if self.image_info.device_handle == protocol_db::INVALID_HANDLE {
-            Some(file_path)
-        } else if file_path.is_null()
+    /// The file path is the full device path for this image and is what is set for the loaded_image_device_path protocol.
+    fn set_file_path(&mut self, file_path: NonNull<efi::protocols::device_path::Protocol>) -> Result<(), EfiError> {
+        let mut fp = file_path.as_ptr();
+
+        // If the device handle is valid, and the handle has a device path protocol, we need to adjust our file path
+        if self.image_info.device_handle != protocol_db::INVALID_HANDLE
             && let Ok(device_path) = PROTOCOL_DB
                 .get_interface_for_handle(self.image_info.device_handle, efi::protocols::device_path::PROTOCOL_GUID)
         {
-            // strip the parent device path prefix from the full device path to leave only the file node
             let (_, device_path_size) =
                 device_path_node_count(device_path as *mut efi::protocols::device_path::Protocol)?;
             let split_idx =
```

Comment:
nit: might be worth a clarifying comment that the subtract here is removing the device path end node.

Reply:
Hi John. I just wrote a test regarding this, and I do not see the END node being subtracted. This is the test:

1. Device handle with path `PCI(0,1C)/PCI(0,0)/END`
2. child full path: `PCI(0,1C)/PCI(0,0)/EFI/BOOT/BOOTX64.EFI/END`
3. Final device path: `EFI/BOOT/BOOTX64.EFI/END` < device path end node still here

Is this expected? Do I need to make some logical changes?

I'll be pushing the test in just a second, it is called `test_set_file_path_with_a_device_handle`